### PR TITLE
Enhanced daemons flags

### PR DESCRIPTION
**Pull request by aanm**
Reordered daemons flags in `daemon/main.go` and rewrite some of the flags descriptions that could cause some misinterpretation.

Also add the bindata.go to the list of ignored git files.

```
cilium-agent --help
Run cilium daemon

Usage:
  daemon [flags]

Flags:
      --access-log string             Path to access log of all HTTP requests observed
      --agent-labels stringSlice      Additional labels to identify this agent
      --allow-localhost string        Policy when to allow local stack to reach local endpoints { auto | always | policy }  (default "auto")
      --bpf-root string               BPF mount path filesystem
      --config string                 Configuration file (default "$HOME/ciliumd.yaml")
  -D, --debug                         Enable debug messages
  -d, --device string                 Device to snoop on (default "undefined")
      --disable-conntrack             Disable connection tracking
      --disable-ipv4                  Disable IPv4 mode
  -e, --docker string                 Register a listener for docker events on the given endpoint (default "unix:///var/run/docker.sock")
      --enable-policy string          Enable policy enforcement (default "default")
      --enable-tracing                Enable tracing while determining policy
      --ipv4-range string             IPv4 prefix for endpoints, must be in correct format (10.N.0.1/16)
      --k8s-api-server string         Kubernetes api address server (for https use --k8s-kubeconfig-path instead)
      --k8s-kubeconfig-path string    Absolute path of the kubernetes kubeconfig file
      --keep-config                   When restoring state, keeps containers' configuration in place
      --keep-templates                Do not restore BPF template files from binary
      --kvstore string                Key-value store type
      --kvstore-opt map               Key-value store options (default map[])
      --label-prefix-file string      Valid label prefixes file path
      --labels stringSlice            List of label prefixes used to determine identity of an endpoint
      --lb string                     Enables load balancer mode where load balancer bpf program is attached to the given interface
      --lib-dir string                Directory path to store runtime build environment (default "/var/lib/cilium")
      --log-driver stringSlice        Logging endpoints to use for example systemd, fluentd
      --log-opt map                   Log driver options for cilium (default map[])
      --logstash                      Enable logstash integration
      --logstash-agent string         Logstash agent address (default "127.0.0.1:8080")
      --logstash-probe-timer uint32   Logstash probe timer (seconds) (default 10)
      --nat46-range string            IPv6 prefix to map IPv4 addresses to (default "0:0:0:0:0:FFFF::/96")
  -n, --node-address string           IPv6 address of node, must be in correct format (XXXX:XXXX:XXXX:XXXX:XXXX:XXXX::/96)
      --restore                       Restores state, if possible, from previous daemon
      --socket-path string            Sets daemon's socket path to listen for connections (default "/var/run/cilium/cilium.sock")
      --state-dir string              Directory path to store runtime state (default "/var/run/cilium")
  -t, --tunnel string                 Tunnel mode vxlan or geneve (default "vxlan")
      --version                       Print version information
```

### CODE REVIEW

**Review by aalemayhu — APPROVED** [review]

File: daemon/main.go [hunk 1]
```diff
@@ -219,50 +219,78 @@ func checkMinRequirements() {
...
+	flags.StringVarP(&v6Address,
+		"node-address", "n", "", "IPv6 address of node, must be in correct format (XXXX:XXXX:XXXX:XXXX:XXXX:XXXX::/96)")
+	flags.BoolVar(&config.RestoreState,
+		"restore", false, "Restores state, if possible, from previous daemon")
+	flags.StringVar(&socketPath,
+		"socket-path", defaults.SockPath, "Sets daemon's socket path to listen for connections")
+	flags.StringVar(&config.RunDir,
+		"state-dir", defaults.RuntimePath, "Directory path to store runtime state")
+	flags.StringVarP(&config.Tunnel,
+		"tunnel", "t", "vxlan", "Tunnel mode VXLAN or GENEVE")
```
Comment (borkmann) [inline daemon/main.go]: This one also has other mode options like "direct" etc, no?

File: daemon/main.go [hunk 2]
```diff
@@ -219,50 +219,78 @@ func checkMinRequirements() {
...
+	flags.StringVar(&labelPrefixFile,
+		"label-prefix-file", "", "Valid label prefixes file path")
+	flags.StringSliceVar(&validLabels,
+		"labels", []string{}, "List of label prefixes used to determine identity of an endpoint")
+	flags.StringVar(&config.LBInterface,
+		"lb", "", "Enables load balancer mode where load balancer bpf program is attached to the given interface")
+	flags.StringVar(&config.LibDir,
+		"lib-dir", defaults.LibraryPath, "Directory path to store runtime build environment")
+	flags.StringSliceVar(&loggers,
+		"log-driver", []string{}, "Logging endpoints to use for example systemd, fluentd")
```
Comment (ianvernon) [inline daemon/main.go]: systemd -> syslog

File: daemon/main.go [hunk 3]
```diff
@@ -219,50 +219,78 @@ func checkMinRequirements() {
...
+	flags.StringVarP(&config.DockerEndpoint,
+		"docker", "e", "unix:///var/run/docker.sock", "Register a listener for docker events on the given endpoint")
+	flags.StringVar(&config.EnablePolicy,
+		"enable-policy", endpoint.DefaultEnforcement, "Enable policy enforcement")
+	flags.BoolVar(&enableTracing,
+		"enable-tracing", false, "Enable tracing while determining policy")
+	flags.StringVar(&v4Prefix,
+		"ipv4-range", "", "IPv4 prefix for endpoints, must be in correct format (10.N.0.1/16)")
+	flags.StringVar(&config.K8sEndpoint,
+		"k8s-api-server", "", "Kubernetes api address server (for https use kubeconfig instead)")
```
Comment (ianvernon) [inline daemon/main.go]: I'd change kubeconfig --> "k8s-kubeconfig-path" so that it corresponds to the flag the user should provide.

Comment (aanm) [reply daemon/main.go]: cool I like it! fixed

File: daemon/main.go (hunk 2 above)
Comment (aanm) [reply daemon/main.go]: for some reason I read systemd, fixed

File: daemon/main.go (hunk 1 above)
Comment (aanm) [reply daemon/main.go]: I believe "direct" mode is only set if the `snoop-device != "undefined"`.
For example, if you set `snoop-device="eth0"` and `-t vxlan` the `"vxlan"` is ignored.
The code is this one: https://github.com/cilium/cilium/blob/9c6885a3a4b81c2355c6981b4126f3a58c3eae1d/daemon/daemon.go#L315

**Review by ianvernon — APPROVED** [review]

File: daemon/main.go [hunk 4]
```diff
@@ -219,50 +219,77 @@ func checkMinRequirements() {
...
-	flags.StringSliceVar(&loggers, "log-driver", []string{}, "logging endpoints to use")
-	flags.Var(option.NewNamedMapOptions("log-opts", &logOpts, nil), "log-opt", "log driver options for cilium")
+	flags.String(
+		"access-log", "", "Path to access log of all HTTP requests observed")
+	flags.StringSlice(
+		"agent-labels", []string{}, "Additional labels to identify this agent")
+	flags.StringVar(&config.AllowLocalhost,
+		"allow-localhost", AllowLocalhostAuto, "Policy when to allow local stack to reach local endpoints { auto | always | policy } ")
+	flags.StringVar(&bpfRoot,
+		"bpf-root", "", "BPF mount path filesystem")
```
Comment (tgraf) [inline daemon/main.go]: `"Path to BPF filesystem"`

File: daemon/main.go [hunk 5]
```diff
@@ -219,50 +219,77 @@ func checkMinRequirements() {
...
+	flags.StringVar(&cfgFile,
+		"config", "", `Configuration file (default "$HOME/ciliumd.yaml")`)
+	flags.BoolP(
+		"debug", "D", false, "Enable debug messages")
```
Comment (tgraf) [inline daemon/main.go]: `"Enable debugging mode"`

File: daemon/main.go [hunk 6]
```diff
@@ -219,50 +219,77 @@ func checkMinRequirements() {
...
+	flags.StringVarP(&config.Device,
+		"device", "d", "undefined", "Device to snoop on")
```
Comment (tgraf) [inline daemon/main.go]: `"Device facing cluster/external network for direct L3 (non-overlay mode)"`

File: daemon/main.go [hunk 7]
```diff
@@ -219,50 +219,77 @@ func checkMinRequirements() {
...
+	flags.BoolVar(&disableConntrack,
+		"disable-conntrack", false, "Disable connection tracking")
+	flags.BoolVar(&config.IPv4Disabled,
+		"disable-ipv4", false, "Disable IPv4 mode")
+	flags.StringVarP(&config.DockerEndpoint,
+		"docker", "e", "unix:///var/run/docker.sock", "Register a listener for docker events on the given endpoint")
```
Comment (tgraf) [inline daemon/main.go]: `"Path to docker runtime socket"`

File: daemon/main.go [hunk 8]
```diff
@@ -219,50 +219,77 @@ func checkMinRequirements() {
...
+	flags.StringVar(&config.EnablePolicy,
+		"enable-policy", endpoint.DefaultEnforcement, "Enable policy enforcement")
+	flags.BoolVar(&enableTracing,
+		"enable-tracing", false, "Enable tracing while determining policy")
```
Comment (tgraf) [inline daemon/main.go]: `"Enable tracing while determining policy (debugging)"`

File: daemon/main.go [hunk 9]
```diff
@@ -219,50 +219,77 @@ func checkMinRequirements() {
...
+	flags.StringVar(&v4Prefix,
+		"ipv4-range", "", "IPv4 prefix for endpoints, must be in correct format (10.N.0.1/16)")
```
Comment (tgraf) [inline daemon/main.go]: `"IPv4 allocation range for local endpoints"`

File: daemon/main.go [hunk 10]
```diff
@@ -219,50 +219,77 @@ func checkMinRequirements() {
...
+	flags.StringVar(&config.K8sEndpoint,
+		"k8s-api-server", "", "Kubernetes api address server (for https use --k8s-kubeconfig-path instead)")
+	flags.StringVar(&config.K8sCfgPath,
+		"k8s-kubeconfig-path", "", "Absolute path of the kubernetes kubeconfig file")
+	flags.BoolVar(&config.KeepConfig,
+		"keep-config", false, "When restoring state, keeps containers' configuration in place")
+	flags.BoolVar(&config.KeepTemplates,
+		"keep-templates", false, "Do not restore BPF template files from binary")
```
Comment (tgraf) [inline daemon/main.go]: We should rename this option to `--keep-bpf-templates`

**Review by tgraf — CHANGES_REQUESTED** [review]
Some more suggestions

File: daemon/main.go (hunk 4 above)
Comment (aanm) [reply daemon/main.go]: Wouldn't this be clear if it says the word "mounted", like "... mounted filesystem"?

File: daemon/main.go (hunk 10 above)
Comment (aanm) [reply daemon/main.go]: Should I create a new flag `--keep-bpf-templates` and leave the `--keep-templates` one for 2 versions with `(deprecated)` written in the description?

Comment (tgraf) [reply daemon/main.go]: Just renamed it. If somebody complains we can add a deprecated version. We haven't declared our CLI stable yet.

File: daemon/main.go (hunk 4 above)
Comment (tgraf) [reply daemon/main.go]: It's not necessarily mounted yet. We will mount the bpf filesystem.

Comment (aanm) [reply daemon/main.go]: Ah that's right!

**Review by tgraf — APPROVED** [review]

**Pull request merged**